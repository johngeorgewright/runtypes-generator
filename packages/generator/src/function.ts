import { SyntaxKind, Type } from 'ts-morph'

export function getFunctionName(type: Type) {
  const symbol = type.getAliasSymbol() || type.getSymbolOrThrow()
  const firstDeclaration = symbol.getDeclarations()[0]
  return firstDeclaration.isKind(SyntaxKind.ArrowFunction)
    ? firstDeclaration
        .getFirstAncestorByKind(SyntaxKind.VariableDeclaration)
        ?.getFirstChildByKind(SyntaxKind.Identifier)
        ?.getText()
    : symbol.getName()
}
