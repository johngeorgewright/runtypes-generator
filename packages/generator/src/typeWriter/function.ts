import { FunctionDeclaration, Node, Signature, Type } from 'ts-morph'
import { last } from '../util'
import generateOrReuseType from './generateOrReuseType'
import TypeWriter from './TypeWriter'
import { Import, ImportFromSource, Static, Write } from './symbols'

export default function* functionTypeWriter(
  type: Type,
  name?: string
): TypeWriter {
  const signature = last(type.getCallSignatures())
  name =
    name ||
    type.getAliasSymbol()?.getName() ||
    type.getSymbolOrThrow().getName()

  const contract = isAsync(signature) ? 'AsyncContract' : 'Contract'
  const importAlias = `_${name}`

  yield [ImportFromSource, { name, alias: importAlias }]
  yield [Import, contract]
  yield [Write, `${contract}(`]

  for (const param of signature.getParameters()) {
    const paramDec = last(param.getDeclarations())
    yield* generateOrReuseType(paramDec.getType())
    if (!paramDec.getType().isNullable() && isOptionalParam(paramDec))
      yield [Write, '.optional()']
    yield [Write, ',']
  }

  yield* generateOrReuseType(
    isAsync(signature)
      ? signature.getReturnType().getTypeArguments()[0]
      : signature.getReturnType()
  )

  yield [Write, ')']

  if (isFunctionDeclaration(type)) {
    yield [Write, `.enforce(${importAlias})`]
    yield [Static, `typeof ${name}`]
  } else yield [Static, importAlias]
}

function isFunctionDeclaration(type: Type) {
  return type
    .getSymbol()
    ?.getDeclarations()
    .some((d) => d instanceof FunctionDeclaration)
}

function isAsync(signature: Signature) {
  return signature.getReturnType().getTargetType()?.getText() === 'Promise<T>'
}

function isOptionalParam(node: Node) {
  return node.getText().includes('?: ')
}
