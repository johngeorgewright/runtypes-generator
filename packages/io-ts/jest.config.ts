import { InitialOptionsTsJest } from 'ts-jest'

const config: InitialOptionsTsJest = {
  globals: {
    'ts-jest': {
      tsconfig: 'tsconfig.test.json',
    },
  },
  preset: 'ts-jest',
  testEnvironment: 'node',
}

export default config
