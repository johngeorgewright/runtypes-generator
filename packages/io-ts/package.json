{
  "name": "@runtyping/io-ts",
  "version": "1.3.1",
  "description": "Generate io-ts from static types & JSON schema.",
  "main": "dist/index.js",
  "scripts": {
    "build": "yarn clean && tsc",
    "clean": "rimraf dist",
    "start": "tsc --watch --preserveWatchOutput",
    "test": "jest"
  },
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/johngeorgewright/runtyping/issues"
  },
  "homepage": "https://github.com/johngeorgewright/runtyping#readme",
  "bin": {
    "runtyping": "dist/cli.js"
  },
  "peerDependencies": {
    "io-ts": "^2.0.0"
  },
  "devDependencies": {
    "@runtyping/test-type-writers": "workspace:*",
    "@semantic-release/commit-analyzer": "9.0.2",
    "@semantic-release/git": "10.0.1",
    "@semantic-release/github": "8.0.6",
    "@semantic-release/npm": "9.0.1",
    "@semantic-release/release-notes-generator": "10.0.3",
    "@types/jest": "29.2.0",
    "@types/json-schema": "7.0.11",
    "@types/node": "16.11.66",
    "fp-ts": "2.13.1",
    "io-ts": "2.2.19",
    "jest": "29.2.0",
    "rimraf": "3.0.2",
    "semantic-release": "19.0.5",
    "semantic-release-monorepo": "7.0.5",
    "ts-jest": "29.0.3",
    "ts-node": "10.9.1",
    "typescript": "4.8.4"
  },
  "dependencies": {
    "@johngw/array": "3.3.0",
    "@runtyping/generator": "workspace:*",
    "@ts-morph/common": "0.17.0",
    "title-case": "3.0.3",
    "ts-morph": "16.0.0",
    "tslib": "2.4.0"
  }
}
