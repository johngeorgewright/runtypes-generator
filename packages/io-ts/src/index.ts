export { default as Generator } from './Generator'
export * from '@runtyping/generator/dist/runtypes'
export * as validators from './validators'
