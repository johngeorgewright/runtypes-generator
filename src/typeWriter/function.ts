import { Type } from 'ts-morph'
import { last } from '../util'
import generateOrReuseType from './generateOrReuseType'
import RuntypeGenerator from './RuntypeGenerator'
import { Import, ImportFromSource, Write } from './symbols'

export default function* functionTypeGenerator(type: Type): RuntypeGenerator {
  const signature = last(type.getCallSignatures())
  const name =
    type.getAliasSymbol()?.getName() || type.getSymbolOrThrow().getName()

  yield [ImportFromSource, name]
  yield [Import, 'Contract']
  yield [Write, 'Contract(']

  for (const param of signature.getParameters()) {
    yield* generateOrReuseType(last(param.getDeclarations()).getType())
    yield [Write, ',']
  }

  yield* generateOrReuseType(signature.getReturnType())

  yield [Write, `).enforce(_${name})`]
}
