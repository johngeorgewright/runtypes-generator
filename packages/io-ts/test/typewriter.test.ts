import testTypeWriter from '@runtyping/test-type-writers'
import * as t from 'io-ts'
import IoTsTypeWriters from '../src/IoTsTypeWriters'
import { fold } from 'fp-ts/lib/Either'
import { pipe } from 'fp-ts/lib/function'

testTypeWriter<t.Type<any>>({
  createNumberValidator: () => t.number,
  createStringValidator: () => t.string,
  createObjectValidator: t.type,

  // Tuples are broken in io-ts. https://github.com/gcanti/io-ts/issues/503
  ignore: [
    'tuple.A',
    'tuple.B',
    'maxItems.schema.json.ExampleSchema',
    'minItems.schema.json.ExampleSchema',
    'variadicTuples.C',
  ],

  typeWriters: new IoTsTypeWriters(),

  validate(validator, data) {
    pipe(
      validator.decode(data),
      fold(
        (errors) => {
          throw new Error(errors.map((error) => error.message).join('\n'))
        },
        () => {}
      )
    )
  },
})
