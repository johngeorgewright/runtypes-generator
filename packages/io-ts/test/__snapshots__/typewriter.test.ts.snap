// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`without static types 1`] = `
"import { type, literal, string, number, union } from 'io-ts';

export const One = type({ type: literal(\\"one\\"), value: string, });
export const Two = type({ type: literal(\\"two\\"), value: number, });
export const Three = union([One, Two,]);
"
`;
